pkg/service/v1alpha1/node.go

```diff
@@ -141,59 +166,68 @@ func (ns *node) NodePublishVolume(
 
 	err = addVolumeToTransitionList(volumeID, apis.CSIVolumeStatusMountUnderProgress)
 	if err != nil {
-		goto PublishVolumeResponse
+		return nil, status.Error(codes.Internal, err.Error())
 	}
 	defer removeVolumeFromTransitionList(volumeID)
 
-	vol, err = prepareVolSpecAndWaitForVolumeReady(req, nodeID)
-	if err != nil {
-		goto PublishVolumeResponse
+	if vol, err = prepareVolSpecAndWaitForVolumeReady(req, nodeID); err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
 	}
-	currentMounts, err = utils.GetMounts(volumeID)
+	isMountRequired, err = VerifyAndUpdateMount(volumeID, targetPath)
 	if err != nil {
-		goto PublishVolumeResponse
-	} else if len(currentMounts) == 1 {
-		if currentMounts[0] == targetPath {
-			goto PublishVolumeResponse
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	if isMountRequired {
+		// Permission is changed for the local directory before the volume is
+		// mounted on the node. This helps to resolve cases when the CSI driver
+		// Unmounts the volume to remount again in required mount mode(ro/rw),
+		// the app starts writing directly in the local directory.
+		// As soon as the volume is mounted the permissions of this directory are
+		// automatically changed to allow Reads and writes.
+		// And as soon as it is unmounted permissions change
+		// back to what we are setting over here.
+		if err = utils.ChmodMountPath(vol.Spec.Volume.MountPath); err != nil {
+			return nil, status.Error(codes.Internal, err.Error())
 		}
-		if err = iscsi.Unmount(currentMounts[0]); err != nil {
-			goto PublishVolumeResponse
+		// Login to the volume and attempt mount operation on the requested path
+		if devicePath, err = iscsi.AttachAndMountDisk(vol); err != nil {
+			return nil, status.Error(codes.Internal, err.Error())
 		}
-	} else if len(currentMounts) > 1 {
-		logrus.Fatalf(
-			"More than one mounts for volume:%s mounts: %v",
-			volumeID, currentMounts,
-		)
+		// TODO update status also in below function
```

Comment:
it seems - we hadn't taken care of this TODO.. i.e., are we updating status now?